contrib/test.sh

```diff
@@ -0,0 +1,130 @@
+#!/bin/sh
+
+set -ex
+
+FEATURES="bitcoin elements test-utils serde"
+
+if [ "$DO_COV" = true ]
+then
+    export RUSTFLAGS="-C link-dead-code"
+fi
+
+cargo --version
+rustc --version
+
+# Some tests require certain toolchain types.
+NIGHTLY=false
+STABLE=true
+if cargo --version | grep nightly; then
+    STABLE=false
+    NIGHTLY=true
+fi
+if cargo --version | grep beta; then
+    STABLE=false
+fi
+
+# Pin dependencies as required if we are using MSRV toolchain.
+if cargo --version | grep "1\.48"; then
+    # 1.0.100 uses syn 2.0 which uses edition 2021
+    cargo update -p serde_json --precise 1.0.99
+    # 1.0.157 uses syn 2.0
+    cargo update -p serde --precise 1.0.156
+    # 1.0.31 uses proc-macro 1.0.66
+    cargo update -p quote --precise 1.0.30
+    # 1.0.66 uses edition 2021
+    cargo update -p proc-macro2 --precise 1.0.65
+fi
+
+if [ "$DO_LINT" = true ]
+then
+    cargo clippy --locked --all-features --all-targets -- -D warnings
+
+    # We should not have any duplicate dependencies. This catches mistakes made upgrading dependencies
+    # in one crate and not in another (e.g. upgrade bitcoin_hashes in bitcoin but not in secp).
+    duplicate_dependencies=$(
+        # Only show the actual duplicated deps, not their reverse tree, then
+        # whitelist the 'syn' crate which is duplicated but it's not our fault.
+        cargo tree  --target=all --all-features --duplicates \
+            | grep '^[0-9A-Za-z]' \
+            | grep -v 'syn' \
+            | wc -l
+                          )
+    if [ "$duplicate_dependencies" -ne 0 ]; then
+        echo "Dependency tree is broken, contains duplicates"
+        cargo tree  --target=all --all-features --duplicates
+        exit 1
+    fi
+fi
+
+echo "********* Testing no features *************"
+# Test without any features
+cargo test --locked --verbose --no-default-features
+
+echo "********* Testing default *************"
+# Then test with the default features
+cargo test --verbose
```

Comment:
In b2aabc679853cb221b7705487b5952b3f055235f:

Let's add an `cargo test --all-features` here as well. We should also do `cargo build` before every `cargo test`, since building does a slightly different compilation.